python/adjoint/Objective.py

```diff
@@ -0,0 +1,661 @@
+            df_drqn = sympy.diff(self.fexpr,self.riqsymbols[2*nq+0])
+            df_diqn = sympy.diff(self.fexpr,self.riqsymbols[2*nq+1])
+            self.dfexpr.append( df_drqn - sympy.I*df_diqn )
+
+
+  ######################################################################
+  ######################################################################
+  ######################################################################
+    def get_fq(self, DFTCells, nf=0):
+
+        # fetch updated values for all objective quantities
+        for nq, qr in enumerate(self.qrules):
+            self.qvals[nq] = DFTCells[qr.ncell].eval_quantity(qr.code,qr.mode,nf)
+            self.riqvals[self.riqsymbols[2*nq+0]]=np.real(self.qvals[nq])
+            self.riqvals[self.riqsymbols[2*nq+1]]=np.imag(self.qvals[nq])
+
+        # plug in objective-quantity values to get value of objective function
+        fval=self.fexpr.evalf(subs=self.riqvals)
+        fval=complex(fval) if fval.is_complex else float(fval)
+
+        return np.array( [fval] + list(self.qvals) )
+
+    # compute values of all partial derivatives \partial f / \partial q
+    def get_partials(self):
+        return np.array( [ df.evalf(subs=self.riqvals) for df in self.dfexpr ] )
+
+#########################################################
+# end of ObjectiveFunction class definition
+#########################################################
+
+#########################################################
+#########################################################
+#########################################################
+class AdjointSolver(object):
+
+    #########################################################
+    #########################################################
+    #########################################################
+    def __init__(self, obj_func, dft_cells, basis, sim=None, vis=None):
+
+        self.obj_func    = obj_func
+        self.dft_cells   = dft_cells
+        self.basis       = basis
+        self.sim         = sim
+        self.vis         = vis
+        self.dfdEps      = None #0.0j*np.zeros(self.dft_cells[-1].slice_dims)
+
+        # prefetch names of outputs computed by forward and
+        # adjoint solves, for use in writing log files
+        self.fqnames     = ['f'] + obj_func.qnames
+        self.bnames      = basis.names
+
+    #########################################################
+    #########################################################
+    #########################################################
+    def eval_fq(self, nf=0):
+        return self.obj_func.get_fq(self.dft_cells,nf=nf)
+
+    def eval_gradf(self, nf=0):
+        cell=self.dft_cells[-1] # design cell
+        EH_forward=cell.get_EH_slices(nf,label='forward')
+        EH_adjoint=cell.get_EH_slices(nf) # no label->current simulation
+        self.dfdEps=np.sum( [EH_forward[nc]*EH_adjoint[nc]
+                               for nc,c in enumerate(cell.components) if c in Exyz], 0 )
```

Comment:
Doesn't a comprehension like this produce a Python `list` and not a numpy array?  On a Python list, my recollection is that you might as well do `sum` … `np.sum` offers no advantage and may actually be slower.  Not that it matters here since the dominant cost will be building the list in the first place.